### PR TITLE
fix:remediations chainlight 001, 002, 004

### PR DESCRIPTION
**Pull request by livingrockrises**

### CODE REVIEW

File: contracts/sponsorship/BiconomySponsorshipPaymaster.sol [hunk 1]
```diff
@@ -47,7 +47,8 @@ contract BiconomySponsorshipPaymaster is
     // Offset in PaymasterAndData to get to PAYMASTER_ID_OFFSET
     uint256 private constant PAYMASTER_ID_OFFSET = PAYMASTER_DATA_OFFSET;
```
Comment (filmakarov) [inline contracts/sponsorship/BiconomySponsorshipPaymaster.sol:49]: Just out of curiosity, any specific reason to do that?

File: contracts/sponsorship/BiconomySponsorshipPaymaster.sol [hunk 2]
```diff
@@ -309,6 +313,14 @@ contract BiconomySponsorshipPaymaster is
...
 
+        // Note: may not need to cache and forward in the context
+        uint256 unaccountedGasCatched = unaccountedGas;
+
+
+        if(unaccountedGasCatched >= userOp.unpackPostOpGasLimit()) {
+            revert PostOpGasLimitTooLow();
+        } 
+
         bool validSig = verifyingSigner.isValidSignatureNow(
```
Comment (filmakarov) [inline contracts/sponsorship/BiconomySponsorshipPaymaster.sol]: Can external signer be SA?
-  if yes, suggest using `isValidSignatureNowCalldata` 
- if no, we should probably switch to ecrecover

File: contracts/sponsorship/BiconomySponsorshipPaymaster.sol [hunk 3]
```diff
@@ -256,26 +257,30 @@ contract BiconomySponsorshipPaymaster is
...
-
-            if (adjustedGasCost > actualGasCost) {
-                // Apply priceMarkup to fee collector balance
-                uint256 premium = adjustedGasCost - actualGasCost;
-                paymasterIdBalances[feeCollector] += premium;
-                // Review if we should emit adjustedGasCost as well
-                emit PriceMarkupCollected(paymasterId, premium);
+            if(prechargedAmount - adjustedGasCost > 0) {
+            // If overcharged refund the excess
+            paymasterIdBalances[paymasterId] += (prechargedAmount -adjustedGasCost);
```
Comment (filmakarov) [inline contracts/sponsorship/BiconomySponsorshipPaymaster.sol]:
```suggestion
            if (prechargedAmount > adjustedGasCost) {
                // If overcharged refund the excess
                paymasterIdBalances[paymasterId] += (prechargedAmount - adjustedGasCost);
```

**Review by filmakarov — CHANGES_REQUESTED** [review]

File: contracts/sponsorship/BiconomySponsorshipPaymaster.sol (hunk 2 above)
Comment (livingrockrises) [reply contracts/sponsorship/BiconomySponsorshipPaymaster.sol]: it's always an EOA. can do plain ecrecover

File: contracts/sponsorship/BiconomySponsorshipPaymaster.sol (hunk 1 above)
Comment (livingrockrises) [reply contracts/sponsorship/BiconomySponsorshipPaymaster.sol:49]: just for readability
PAYMASTER_DATA_OFFSET is always fixed.
PAYMASTER_ID_OFFSET depends on design how our paymasterandData is structured.

**Comment by livingrockrises** [comment]
can we also check CI build issue in this PR itself.
I also don't think we can tackle coverage at this point because this repo only has foundry and we must use viIR.

**Comment by filmakarov** [comment]
> can we also check CI build issue in this PR itself. 

The build issue for me was resolved with downgrading yarn to 1.2.xx

> I also don't think we can tackle coverage at this point because this repo only has foundry and we must use viIR.

why do we must use via-IR?

**Comment by filmakarov** [comment]
> can we also check CI build issue in this PR itself. 

The build issue for me was resolved with downgrading yarn to 1.2.xx

> I also don't think we can tackle coverage at this point because this repo only has foundry and we must use viIR.

why do we must use via-IR?

**Comment by livingrockrises** [comment]
> isValidSignatureNow

check getHash. stack too deep.

**Comment by filmakarov** [comment]
> check getHash. stack too deep.

getHash can be fixed.

```
userOp.nonce,
                keccak256(userOp.initCode),
                keccak256(userOp.callData),
                userOp.accountGasLimits,
             uint256(bytes32(userOp.paymasterAndData[PAYMASTER_VALIDATION_GAS_OFFSET:PAYMASTER_DATA_OFFSET])),
                userOp.preVerificationGas,
                userOp.gasFees,
```
this can be packed in one bytes

however, there are further stack too deep errors in the token paymaster

So, the best way to handle is would be this 

`forge coverage --ir-minimum` => [here](https://github.com/foundry-rs/foundry/issues/3527#issuecomment-2271987183)

File: contracts/sponsorship/BiconomySponsorshipPaymaster.sol (hunk 2 above)
Comment (livingrockrises) [reply contracts/sponsorship/BiconomySponsorshipPaymaster.sol]: done

**Comment by livingrockrises** [comment]
> > check getHash. stack too deep.
> 
> getHash can be fixed.
> 
> ```
> userOp.nonce,
>                 keccak256(userOp.initCode),
>                 keccak256(userOp.callData),
>                 userOp.accountGasLimits,
>              uint256(bytes32(userOp.paymasterAndData[PAYMASTER_VALIDATION_GAS_OFFSET:PAYMASTER_DATA_OFFSET])),
>                 userOp.preVerificationGas,
>                 userOp.gasFees,
> ```
> 
> this can be packed in one bytes
> 
> however, there are further stack too deep errors in the token paymaster
> 
> So, the best way to handle is would be this
> 
> `forge coverage --ir-minimum` => [here](https://github.com/foundry-rs/foundry/issues/3527#issuecomment-2271987183)

will use --ir-minimum then.
however we can change getHash code if we want.

**Comment by livingrockrises** [comment]
> > > check getHash. stack too deep.
> > 
> > 
> > getHash can be fixed.
> > ```
> > userOp.nonce,
> >                 keccak256(userOp.initCode),
> >                 keccak256(userOp.callData),
> >                 userOp.accountGasLimits,
> >              uint256(bytes32(userOp.paymasterAndData[PAYMASTER_VALIDATION_GAS_OFFSET:PAYMASTER_DATA_OFFSET])),
> >                 userOp.preVerificationGas,
> >                 userOp.gasFees,
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > this can be packed in one bytes
> > however, there are further stack too deep errors in the token paymaster
> > So, the best way to handle is would be this
> > `forge coverage --ir-minimum` => [here](https://github.com/foundry-rs/foundry/issues/3527#issuecomment-2271987183)
> 
> will use --ir-minimum then. however we can change getHash code if we want.

I will fix this in another PR to revamp test cases.

**Comment by livingrockrises** [comment]
closing in favour of https://github.com/bcnmy/gasdaddy/pull/24

**Pull request closed without merging**